src/HDF5.jl

```diff
@@ -1272,12 +1281,33 @@ Base.getindex(parent::Union{File,Group}, r::Reference) = _deref(parent, r)
 Base.getindex(parent::Dataset, r::Reference) = _deref(parent, r) # defined separately to resolve ambiguity
 
 # convert special types to native julia types
-normalize_types(x) = x
-normalize_types(x::NamedTuple{T}) where {T} = NamedTuple{T}(map(normalize_types, values(x)))
-normalize_types(x::Cstring) = unsafe_string(x)
-normalize_types(x::FixedString) = unpad(String(collect(x.data)), pad(x))
-normalize_types(x::FixedArray) = normalize_types.(reshape(collect(x.data), size(x)...))
-normalize_types(x::VariableArray) = normalize_types.(copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false)))
+function normalize_types(::Type{T}, buf::AbstractMatrix{UInt8}) where {T}
+    return [_normalize_types(T, view(buf, :, ind)) for ind in axes(buf, 2)]
+end
+function _typed_load(::Type{T}, buf::AbstractVector{UInt8}) where {T}
+    return @inbounds reinterpret(T, buf)[1]
+end
+
+_normalize_types(::Type{T}, buf::AbstractVector{UInt8}) where {T} = _typed_load(T, buf)
+function _normalize_types(::Type{T}, buf::AbstractVector{UInt8}) where {K, T <: NamedTuple{K}}
+    nv = ntuple(fieldcount(T)) do ii
+        elT = fieldtype(T, ii)
+        off = fieldoffset(T, ii)
+        sub = view(buf, off .+ (1:sizeof(elT)))
+        return _normalize_types(elT, sub)
+    end
+    return NamedTuple{K}(nv)
+end
+function _normalize_types(::Type{V}, buf::AbstractVector{UInt8}) where {T, V <: VariableArray{T}}
+    va = _typed_load(V, buf)
+    pbuf = unsafe_wrap(Array, convert(Ptr{UInt8}, va.p), (sizeof(T), Int(va.len)), own = false)
+    return normalize_types(T, pbuf)
+end
+function _normalize_types(::Type{F}, buf::AbstractVector{UInt8}) where {T, F <: FixedArray{T}}
+    return reshape(normalize_types(T, reshape(buf, sizeof(T), :)), size(F)...)
+end
+_normalize_types(::Type{Cstring}, buf::AbstractVector{UInt8}) = unsafe_string(reinterpret(Ptr{UInt8}, buf)[1])
```

Comment:
Why does `pointer(buf)` not work in this case?

Reply:
It probably does; I think just a left-over from as I was working my way down to a completely `Array{UInt8}`-based implementation.

Reply:
I'm a bit confused here. 
So on master if we wrote a matrix (say of size 2x2) of fixed strings, the buffer would be 2x2 matrix where each element would be a `Cstring` Ptr. 

Correct me if I'm wrong, in this PR, instead the buffer is a `sizeof(Cstring)` x 4 matrix. I'm lost from this point on how things are read in.

Reply:
A fixed string matrix would have the data inline, yes, but that's read back through the code path `normalize_types` -> `_normalize_types(::Type{<:FixedString}, buf)` (for each element).

This method is called when the stored data is a variable-length string (which is what you get if you try to store a `Matrix` of `String`s):
```julia
julia> datatype(String["a" "b"; "c" "d"])
HDF5.Datatype: H5T_STRING {
      STRSIZE H5T_VARIABLE;
      STRPAD H5T_STR_NULLTERM;
      CSET H5T_CSET_UTF8;
      CTYPE H5T_C_S1;
   }
```

(It looks like we don't have an easy way to construct a fixed-string matrix at the moment:
```julia
julia> fix_strs = let f(x) = HDF5.FixedString{1,0}(tuple(x...))
           [f(b"a") f(b"b"); f(b"c") f(b"d")]
       end
2×2 Matrix{HDF5.FixedString{1, 0}}:
 FixedString{1, 0}((0x61,))  FixedString{1, 0}((0x62,))
 FixedString{1, 0}((0x63,))  FixedString{1, 0}((0x64,))

julia> datatype(fix_strs)
ERROR: MethodError: no method matching datatype(::Matrix{HDF5.FixedString{1, 0}})
Closest candidates are:
  datatype(::HDF5.Dataset) at /home/justin/.julia/dev/HDF5/src/HDF5.jl:988
  datatype(::HDF5.Attribute) at /home/justin/.julia/dev/HDF5/src/HDF5.jl:990
  datatype(::Union{Bool, Float32, Float64, Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8, HDF5.Reference}) at /home/justin/.julia/dev/HDF5/src/HDF5.jl:993
  ...
Stacktrace:
...
```
)

Reply:
@jmert Sorry, but I edited my comment almost the same time you commented.

Reply:
I think I follow the logic for the methods
```julia
_normalize_types(::Type{Cstring}, buf::AbstractVector{UInt8}) = unsafe_string(reinterpret(Ptr{UInt8}, buf)[1])
_normalize_types(::Type{T}, buf::AbstractVector{UInt8}) where {T <: FixedString} = unpad(String(buf), pad(T))
```
I need to go through the other methods, but I think I see how this is operating.

Reply:
> Correct me if I'm wrong, in this PR, instead the buffer is a sizeof(Cstring) x 4 matrix. I'm lost from this point on how things are read in. In this example, is it the case that each column of the buffer is a pointer which in the original example would be the Cstring that would be stored directly? This is then reinterpreted read in and then resized?

Yes, that's principle I'm working with — the first dimension at any stage is bytes for a single data element, whatever the type.

These "leaf" methods (all of the `_normalize_types`) accept only an `AbstractVector{UInt8}` because they're supposed to contain only a single element to interpret.

`normalize_types` works with an `AbstractMatrix` buffer since it will produce an array of outputs, and some of the "leaf"s like for `FixedArray` and `VariableArray` elements re-dispatch through there after reshaping their buffer back to a matrix since they know how many elements they contain.